File: is_subsequence.cpp
/*
392. Is Subsequence

Time Complexity: O(n) where n is length of t
Space Complexity: O(1)
*/

#include <iostream>
#include <string>
using namespace std;


class Solution {
public:
    bool isSubsequence(string s, string t) {
        if (s.empty()) {
            return true;
        }
        
        int i = 0;  // Pointer for s
        
        for (char c : t) {
            if (c == s[i]) {
                i++;
                if (i == s.length()) {
                    return true;
                }
            }
        }
        
        return false;
    }
};




// Test cases
int main() {
    Solution solution;
    
    // Example 1
    cout << "Example 1: " << (solution.isSubsequence("abc", "ahbgdc") ? "true" : "false") << endl;  // Expected: true
    
    // Example 2
    cout << "Example 2: " << (solution.isSubsequence("axc", "ahbgdc") ? "true" : "false") << endl;  // Expected: false
    
    // Edge case: empty s
    cout << "Example 3: " << (solution.isSubsequence("", "ahbgdc") ? "true" : "false") << endl;  // Expected: true
    
    // Edge case: empty t
    cout << "Example 4: " << (solution.isSubsequence("abc", "") ? "true" : "false") << endl;  // Expected: false
    
    // Edge case: both empty
    cout << "Example 5: " << (solution.isSubsequence("", "") ? "true" : "false") << endl;  // Expected: true
    
    // Edge case: s equals t
    cout << "Example 6: " << (solution.isSubsequence("abc", "abc") ? "true" : "false") << endl;  // Expected: true
    
    // Edge case: single character match
    cout << "Example 7: " << (solution.isSubsequence("b", "abc") ? "true" : "false") << endl;  // Expected: true

    return 0;
}

File: is_subsequence.py
"""
392. Is Subsequence

Time Complexity: O(n) where n is length of t
Space Complexity: O(1)
"""

class Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        if not s:
            return True
        
        i = 0
        
        for char in t:
            if char == s[i]:
                i += 1
                if i == len(s):
                    return True
        
        return False

# Test cases
if __name__ == "__main__":
    solution = Solution()
    
    # Example 1
    s1, t1 = "abc", "ahbgdc"
    print(f"Example 1: {solution.isSubsequence(s1, t1)}")  # Expected: True
    
    # Example 2
    s2, t2 = "axc", "ahbgdc"
    print(f"Example 2: {solution.isSubsequence(s2, t2)}")  # Expected: False
    
    # Edge case: empty s
    s3, t3 = "", "ahbgdc"
    print(f"Example 3: {solution.isSubsequence(s3, t3)}")  # Expected: True
    
    # Edge case: empty t
    s4, t4 = "abc", ""
    print(f"Example 4: {solution.isSubsequence(s4, t4)}")  # Expected: False
    
    # Edge case: both empty
    s5, t5 = "", ""
    print(f"Example 5: {solution.isSubsequence(s5, t5)}")  # Expected: True
    
    # Edge case: s equals t
    s6, t6 = "abc", "abc"
    print(f"Example 6: {solution.isSubsequence(s6, t6)}")  # Expected: True
    
    # Edge case: single character match
    s7, t7 = "b", "abc"
    print(f"Example 7: {solution.isSubsequence(s7, t7)}")  # Expected: True
    
    # Edge case: single character no match
    s8, t8 = "d", "abc"
    print(f"Example 8: {solution.isSubsequence(s8, t8)}")  # Expected: False
